Access to System Certificate Store on macOS

reneme opened the pull request:
This is ongoing work towards an adapter for the macOS system certificate store onto Botan's `Certificate_Store` interface. I will add documentation, description and tests along the way and would be happy about a quick code review. Particularly, whether my build-system integration is reasonable. :-)

Currently implemented are:

* `Certificate_Store::find_cert()`
* `Certificate_Store::find_all_certs()`
* `Certificate_Store::find_cert_by_pubkey_sha1()`
* `Certificate_Store::all_subjects()`

However, I'm not sure whether `::find_crl_for()` and `::find_cert_by_raw_subject_dn_sha256()` are in reach.

Furthermore, I'm a bit worried about testing this. The tests added for now try to fetch the root certificate that signs the TLS certificate of "botan.randombit.net" in different ways. However, more sophisticated tests would require root access to the CI machine, to alter the root certificate chains. I don't think that this is feasible.

randombit commented:
Oh nice, thanks.

I don't have time to review this right now but will do so asap, probably in next few days.

We should (I think) have root on the CI so it would be possible to test it there. However we probably should not ask for or require root access to run the tests normally, eg when installing through Homebrew. Nor should we mess with an end users certificate store ;)

To some extent that could be worked around by adding a flag to `botan-test` that signals "I'm running in continuous integration environment and it's ok to run potentially dangerous or destructive tests." and any test modifying the local certstore could check that flag before running.

randombit commented:
Have not had time for a full review but a few quick comments:

- Question: Is it possible to replace `scoped_CFType` with `unique_ptr` which is passed a release function? As done for example in https://github.com/randombit/botan/blob/master/src/lib/prov/openssl/openssl_ec.cpp#L141

- Use `BOTAN_ASSERT` or `BOTAN_ASSERT_EQUAL` instead of C stdlib `assert`. I think `assert_notnull` could be replaced by `BOTAN_ASSERT_NONNULL`.

- Question: is it possible this works on iOS as well?

reneme commented:
Regarding the `scoped_CFType`: I failed to fully replace it with `unique_ptr`. Those `CFTypeRef`-like types are eventually typedefs of `void*`, so they already _are_ a pointer themselves. I tried to replace it like so:

```C++
template <typename T>
struct CFType_deleter {
  void operator()(T cftype) {
    if (cftype != nullptr) {
      CFRelease(cftype);
    }
  }
};

template <typename T>
using scoped_CFType = std::unique_ptr<typename std::remove_pointer<T>::type, CFType_deleter<T>>;
```

This sort-of works, but doesn't allow to take the address of the wrapped `CFTypeRef`. That's needed when using it as output parameter of the `SecKeychainOpen()` call. One could of course add a short-lived bare `CFTypeRef` and wrap it after the call, but I didn't like that approach.

reneme commented:
I replaced the `assert()` calls with `BOTAN_ASSERT` and renamed the `assert_nonnull` and `assert_success` to `check_xxx`, as I intend to throw exceptions in those conditions.

reneme commented:
Regarding iOS: I would imagine that this is mostly going to work on iOS as well. I think, accessing the keychain itself is handled a bit differently there, though. In particular there is only _one_ keychain in the whole system, so the calls to `SecKeychainOpen()` can probably be omitted. A quick glance in the documentation of the used API calls suggests that it should work from iOS 10.6 onwards with minor adaptions. That said, I don't have a test setup readily available to try it.

reneme commented:
Finally, I found a somewhat major bummer that sounds pretty odd to me: In order to successfully find certificates using Apple's `SecItemCopyMatching()` call, I needed to "normalize" the DN objects. See [here](https://github.com/randombit/botan/pull/1830/commits/a27f1c2ef6efb7d90b2f2a44c5e08a13903d7407#diff-ce3705d4ff7e58a182e5ff3f3334b904R78) for the actual method.

Apple states in their [keychain implementation](https://opensource.apple.com/source/Security/Security-55471/sec/Security/SecCertificate.c.auto.html) that ASN.1 strings of the type 'PrintableString' are not case sensitive. Hence, they transform those strings into uppercase and furthermore strip leading and trailing whitespace. Also the certificate lookup seems to be rather slow: The current test body, takes nearly 500ms on my laptop, despite fetching just a handful certificates.

As an alternative of using Apple's APIs for certificate lookups, we might be better off to pull the entire system certificate store into the `Certificate_Store_In_Memory` implementation upon initialization. Initially, I rejected that idea as rather dirty. What do you think?

reneme commented on src/lib/x509/certstor_system_macos/certstor_macos.cpp:
```diff
@@ -0,0 +1,472 @@
+   const auto count = CFArrayGetCount(result.get());
+   BOTAN_ASSERT(count > 0, "certificate result list contains data");
+
+   std::vector<std::shared_ptr<const X509_Certificate>> output;
```
Pre-allocate buffer?
```suggestion
   std::vector<std::shared_ptr<const X509_Certificate>> output;
   output.reserve(count);
```

randombit replied:
It shouldn't matter much since you are just reserving the size for the `shared_ptr`s. Either way is fine as far as I am concerned.

randombit commented:
> Apple states in their [keychain implementation](https://opensource.apple.com/source/Security/Security-55471/sec/Security/SecCertificate.c.auto.html) that ASN.1 strings of the type 'PrintableString' are not case sensitive. Hence, they transform those strings into uppercase and furthermore strip leading and trailing whitespace.

So, this follows from the requirements of RFC 3280 which states that when comparing DNs, "attribute values in PrintableString are not case sensitive" as well as the whitespace normalization. And in fact X.509 proper requires such for all string types, and we implement that when comparing DNs (function `x500_name_cmp` in `parsing.cpp`).

However Apple's behavior here is completely broken: how could "case insensitive" turn out to mean "you must take into account casing"?!? I think they were just too lazy to implement the correct DN comparison algorithm and did this as a hack.

> Also the certificate lookup seems to be rather slow: The current test body, takes nearly 500ms on my laptop, despite fetching just a handful certificates.

Wow that is quite slow. Maybe worth timing each test individually to help diagnose the issue? Call `start_timer`/`end_timer` on your `Test::Result` instances to enable this.

> As an alternative of using Apple's APIs for certificate lookups, we might be better off to pull the entire system certificate store into the `Certificate_Store_In_Memory` implementation upon initialization. Initially, I rejected that idea as rather dirty. What do you think?

Yeah that seems messy, for example consider a long running server process: you would probably expect that adding a cert to the system store would make it available in the server process without a process restart.

I wonder too if perhaps there is some startup overhead to using the system store, say 400ms on first use, but then all subsequent operations are fast. You might be able to detect this using `botan-test --test-runs=10 <test names>` which runs the named tests several times in a row.

reneme commented:
> I wonder too if perhaps there is some startup overhead to using the system store, say 400ms on first use, but then all subsequent operations are fast.

Thanks for the pointer with `start_timer`/`end_timer`. That was helpful. Turns out that the `.all_subjects()` test takes 95% of the full time, i.e. some 450ms. Which is unfortunate, but somewhat expected. It needs to pull in and parse _all_ trusted certificates, after all. Initialization of the `SecKeychain` object is in the ballpark of 1ms. So are individual certificate lookups.

> However Apple's behavior here is completely broken: how could "case insensitive" turn out to mean "you must take into account casing"?!?

Its annoying for sure. I'll have another look whether we can somehow do the "normalization" with some public Apple API, rather than re-implementing it. At least that would be less brittle, I guess.

> you would probably expect that adding a cert to the system store would make it available in the server process without a process restart.

That's exactly what I had in mind as well, when deciding against this option.

randombit commented:
> I'll have another look whether we can somehow do the "normalization" with some public Apple API, rather than re-implementing it. At least that would be less brittle, I guess.

Looks like https://developer.apple.com/documentation/security/1396030-seccertificatecopynormalizedsubj?language=objc is the intended API for this. And yeah, good to use the public API here since they might decide to apply the same normalization to other string types in the future.

reneme commented on src/lib/x509/certstor_system_macos/certstor_macos.cpp:
```diff
@@ -0,0 +1,475 @@
+std::shared_ptr<const X509_CRL> Certificate_Store_MacOS::find_crl_for(const X509_Certificate& subject) const
+   {
+   BOTAN_UNUSED(subject);
+   throw Not_Implemented("Certificate_Store_MacOS::find_crl_for");
```
I'm currently testing this along with our under-construction TLS asio stream (see #1731) and it works quite nicely. However, for the certificate validation to work, I need to statically return an empty list of CRLs (or respectively don't override this function at all, hence falling back to the base class' method returning an empty list). As far as I can see, the macOS keychain does not support any CRL handling. Is "not overriding" a reasonable implementation strategy here?

randombit replied:
Yes if the keychain does not support CRLs at all then it is fine to either omit implementing the function entirely, or (to make it obvious the omission is intentional) override it and have it return an empty list with a comment to the effect that it is not supported by the system API. `Not_Implemented` is more to indicate a TODO type situation, where the functionality is possible but the code is just not written yet.

randombit commented on src/lib/x509/certstor_system_macos/certstor_macos.h:
```diff
@@ -0,0 +1,75 @@
+* opens a handle to the macOS keychain and serves certificate queries directly
+* from there.
+*/
+class BOTAN_PUBLIC_API(2, 0) Certificate_Store_MacOS final : public Certificate_Store
```
This should be 2,10 since that will be the version version that includes this.

randombit commented on src/tests/test_certstor_macos.cpp:
```diff
@@ -0,0 +1,336 @@
+      auto certstore = new Botan::Certificate_Store_MacOS();
+      result.end_timer();
+
+      delete certstore;
```
I know it is just a test but would be better to use RAII here - explicit `delete` is always trouble.

randombit commented on src/build-data/os/darwin.txt:
```diff
@@ -21,6 +21,8 @@ commoncrypto
 threads
 filesystem
+
+certstore
```
Having this feature named certstore is going to get confusing since Windows etc also has a certificate store. Maybe something like `apple_keychain`?

reneme replied:
Judging from "filesystem", "threads" and so on, I assumed that this is a list of "features" a certain platform provides. I.e. if we'd add a similar implementation for Windows (which we are actually planning in the near future), we would just add "certstore" to `windows.txt` as well.
Is that a wrong assumption?

randombit replied:
Here feature is in terms of specific APIs, so for `threads` we mean that `std::thread` and company work, and `sockets` means BSD socket API (which is why Windows defines `winsock2` instead as the APIs differ). If say there was a platform which had threads but the `std::` thread API didn't work and we had to use some system specific API, it would get its own distinct feature and not define `threads`.

reneme replied:
That's good to know, thanks.

randombit commented on src/tests/test_certstor_macos.cpp:
```diff
@@ -0,0 +1,336 @@
+
+#include <iostream>
+
+#if defined(BOTAN_TARGET_OS_IS_DARWIN)
```
This should be checking for `BOTAN_HAS_CERTSTOR_MACOS` since it is possible to build on OSX but with the certstore disabled.

randombit commented on src/lib/x509/certstor_system_macos/certstor_macos.h:
```diff
@@ -0,0 +1,75 @@
+   {
+   public:
+      Certificate_Store_MacOS();
+
```
`=default` copy/move constructors/assignments?

randombit commented on src/lib/x509/certstor_system_macos/certstor_macos.cpp:
```diff
@@ -0,0 +1,475 @@
+
+#include <botan/build.h>
+
+#if defined(BOTAN_TARGET_OS_IS_DARWIN)
```
No need for this - if the file is being compiled we know (by virtue of restrictions set in `info.txt`) that we are on OSX

reneme commented:
> Looks like https://developer.apple.com/documentation/security/1396030-seccertificatecopynormalizedsubj?language=objc is the intended API for this.

Unfortunately this API is deprecated and also takes a `SecCertificateRef`. In [Security/SecCertificate.c](https://opensource.apple.com/source/Security/Security-55471/sec/Security/SecCertificate.c.auto.html) I found an internal API called `SecDistinguishedNameCopyNormalizedContent` (defined in [Security/SecCertificateInternal.h](https://opensource.apple.com/source/Security/Security-55471/sec/Security/SecCertificateInternal.h.auto.html)) which does pretty much what we need. It is not defined in the public header though, so I needed to define it in the compilation unit like so:

```C++
extern "C" CFDataRef SecDistinguishedNameCopyNormalizedContent(CFDataRef distinguished_name);
```

With a bit of further mangling (the call strips the enclosing ASN.1 `SEQUENCE` from the DN data), it actually does what it should. However, I obviously don't feel comfortable linking against internal Apple's internal API... :-/

randombit commented on src/lib/x509/certstor_system_macos/certstor_macos.cpp:
```diff
@@ -0,0 +1,475 @@
+
+   auto chars = CFStringGetLength(cfstring);
+
+   // Worst case: 2 bytes per character + NUL
```
Is that really true? UTF-8 can be up to 4 bytes / character

reneme replied:
According to the [documentation](https://developer.apple.com/documentation/corefoundation/1542853-cfstringgetlength?language=objc) `CFStringGetLength()` returns the number of UTF-16 code pairs. So, for example this Emoji "🧖" requires two UTF-16 code pairs (Big Endian: 0xD83E 0xDDD6) and four UTF-8 bytes (0xF0 0x9F 0xA7 0x96). For a string containing only this emoji `CFStringGetLength()` would return "2", hence the calculation would be correct.

Nevertheless, while reading this up, I found `CFStringGetMaximumSizeForEncoding()` which is probably the better choice anyway.

randombit commented on src/lib/x509/certstor_system_macos/certstor_macos.cpp:
```diff
@@ -0,0 +1,475 @@
+
+   // Worst case: 2 bytes per character + NUL
+   auto cstrlen = chars * 2 + 1;
+   auto cstr = std::unique_ptr<char[]>(new char[cstrlen]);
```
would prefer `vector` to explicit `new`

randombit commented on src/tests/test_certstor_macos.cpp:
```diff
@@ -0,0 +1,336 @@
+
+#include "tests.h"
+
+#include <botan/build.h>
```
You can assume "tests.h" includes `botan/build.h` since most tests want to do feature macro checks.

randombit commented:
> However, I obviously don't feel comfortable linking against internal Apple's internal API...

Yeah that is no good. OK I guess we use the normalization you wrote and hope Apple doesn't change their minds about what is normalized or how.

reneme commented:
Thanks for the review. I fixed the simple things right away and will do the rest tomorrow.

reneme commented:
I'm done with applying your comments.

randombit commented:
There are a few warnings:

```
src/lib/x509/certstor_system_macos/certstor_macos.cpp:245:49: warning: suggest braces around initialization of subobject [-Wmissing-braces]
         std::array<const void*, 2> keychains({ m_system_roots.get(),
                                                ^~~~~~~~~~~~~~~~~~~~~
                                                {
src/lib/x509/certstor_system_macos/certstor_macos.cpp:463:95: warning: unused parameter 'subject' [-Wunused-parameter]
std::shared_ptr<const X509_CRL> Certificate_Store_MacOS::find_crl_for(const X509_Certificate& subject) const
                                                                                              ^
2 warnings generated.
src/tests/test_certstor_macos.cpp:81:44: warning: empty parentheses interpreted as a function declaration [-Wvexing-parse]
      Botan::Certificate_Store_MacOS unused();
                                           ^~
src/tests/test_certstor_macos.cpp:81:44: note: remove parentheses to declare a variable
      Botan::Certificate_Store_MacOS unused();
                                           ^~
src/tests/test_certstor_macos.cpp:81:38: warning: unused function 'unused' [-Wunused-function]
      Botan::Certificate_Store_MacOS unused();
                                     ^
2 warnings generated.
```

Can you fix these and then squash all the commits?

Also, I realized now why all_subjects is slow. In addition to accessing every cert, it has to actually load/parse the certificates to work around Apple's DN mangling. This triggers a signature check on each CA certificate (this is necessary to test self-signed vs self-issued certificates). Likely if that logic were made to be lazy (computed just when/if X509_Certificate::is_self_signed is called) then the operation would be much faster.

reneme commented:
> Can you fix these and then squash all the commits?

Done. ... waiting for CI to see if all the warnings have vanished.

> Also, I realized now why all_subjects is slow. ...

I consider this outside the scope of that PR. Another option would be to let `all_subjects()` throw `Not_Implemented` for the time being. Not sure whether this list is of much value anyway.

randombit commented:
> I consider this outside the scope of that PR.

Oh absolutely - I meant that I will look at addressing that later, so at some point in the future this slow operation will become fast.

reneme commented:
> waiting for CI to see if all the warnings have vanished.

Looks like the code doesn't produce warnings related to this PR on the macOS build anymore. :)

randombit commented:
Merged! Thanks again.
